Firestore/core/src/firebase/firestore/immutable/tree_sorted_map.h

```diff
@@ -97,18 +111,65 @@ class TreeSortedMap : public SortedMapBase, private util::ComparatorHolder<C> {
     return TreeSortedMap{this->comparator()};
   }
 
-  /** Returns true if the map contains no elements. */
-  bool empty() const {
-    return root_.empty();
+  bool contains(const K& key) const {
+    // Inline the tree traversal here to avoid building up the stack required
+    // to construct a full iterator.
+    const C& comparator = this->comparator();
+    const node_type* node = &root();
+    while (!node->empty()) {
+      util::ComparisonResult cmp = util::Compare(key, node->key(), comparator);
+      if (cmp == util::ComparisonResult::Same) {
+        return true;
+      } else if (cmp == util::ComparisonResult::Ascending) {
+        node = &node->left();
+      } else {
+        node = &node->right();
+      }
+    }
+    return false;
   }
 
-  /** Returns the number of items in this map. */
-  size_type size() const {
-    return root_.size();
+  /**
+   * Finds a value in the map.
+   *
+   * @param key The key to look up.
+   * @return An iterator pointing to the entry containing the key, or end() if
+   *     not found.
+   */
+  const_iterator find(const K& key) const {
+    const_iterator found = LowerBound(key);
+    if (!found.is_end() && !this->comparator()(key, found->first)) {
+      return found;
+    } else {
+      return end();
+    }
   }
 
-  const node_type& root() const {
-    return root_;
+  /**
+   * Finds the index of the give key in the map.
+   *
+   * @param key The key to look up.
+   * @return The index of the entry containing the key, or npos if not found.
+   */
+  size_type find_index(const K& key) const {
```

Comment:
For context, how is `find_index` expected to be used?

Reply:
In the computation of the positions of the documents in the changes to the document snapshots. Each change contains the old and new indexes of the documents in the array of documents representing the view.